src/ssl/bio.cc

```diff
@@ -717,6 +728,12 @@ applyTlsDetailsToSSL(SSL *ssl, Security::TlsDetails::Pointer const &details, Ssl
         SSL_set_options(ssl, SSL_OP_NO_COMPRESSION);
 #endif
 
+#if defined(SSL_OP_NO_TLSv1_3)
+    // avoid "inappropriate fallback" OpenSSL error messages
+    if (Security::Tls1p2orEarlier(details->tlsSupportedVersion))
```

Comment:
Should we check that tlsSupportedVersion is _known_ first here?

```suggestion
    if (details->tlsSupportedVersion && Security::Tls1p2orEarlier(details->tlsSupportedVersion))
```

Reply:
It is not really needed here. The tlsSupportedVersion is the first field of details which is parsed while we are parsing TLS hello messages.  Unknown means failed to parse or not parse yet. The details here refers to TLS details of the client-to-squid side. If we failed to parse client TLS then we should not be here.

Reply:
> If we failed to parse client TLS then we should not be here.

The (only) caller almost agrees with your assertion, but not quite:

```C++
if (details && details->tlsVersion.protocol != AnyP::PROTO_NONE)
    applyTlsDetailsToSSL(serverSession.get(), details, csd->sslBumpMode);
```

Note that tlsVersion is not tlsSupportedVersion.

The logic for setting tlsSupportedVersion is very complex and may change at any time. I can imaging a situation where the field is set or even cleared when parsing fails or even when there was no parsing at all (e.g., some kind of "assume the matching client is using TLS v1.3" feature).

Moreover, TLS handshake has many features/fields. For every other feature/field, applyTlsDetailsToSSL() checks that the field was parsed AFAICT. Why make tlsSupportedVersion field special? Even if it should be set now, why risk future bugs instead of following the existing "if set, then apply" pattern in applyTlsDetailsToSSL()?

If you do not object, I will add that check.

Reply:
commit 6dddb61d adds this check.

This check does not hart so I add it. Also we may have such situations in the future, for example we may need to build custom TlsDetails objects to set the desired TLS features..

> Moreover, TLS handshake has many features/fields. For every other feature/field, applyTlsDetailsToSSL() checks that the field was parsed AFAICT

The other fields (probably with the exception of ciphers ) may not be present in a hello message so the check is required for these fields..